#472 Refactor action defintion to interfaces

tortmayr opened the pull request:
Sprotty moved away from class based approach for action definiton and solely  relies on interfaces now.
To be able to update to the latest sprotty version GLSP needs to to do the same. Otherwise action definitions might clash
and there will be an inconsistent mix of class and interface definitions.
This PR refactors the action definitions in the base protocol & the glsp client to interfaces. Action definitions that are reused from sprotty are marked
with `extends sprotty.XYZAction`. In addition, to the pure interface definition for each action/operation type a namespace is provided that offers utility functions for type checking and creating new instances.

- Refactor all action definitions to interfaces+namespaces
- Refactor all action constructions calls to use the new `Action.create() `approach instead of the old `new Action()` approach
- Add tests for the action utility function (is, create etc.) of all protocol actions.
- Replace all old typeguard functions like `isComputedBoundsAction` with their new namespace counterpart (SetBoundsAction.is)
- Add `isOperation`:true boolean property to the operation interface. This serves as discriminator to make actions & operations distinguishable on a
type level.
- Update to latest sprotty version
- Update to latest glsp-config package to  make use of the new no-deprecated eslint rules.
- Add package specific eslint warnings for restricted/discourages imports. 

Also: 
- Disable hover-popup listener when the Ede creation tool is active
- Update changelog

Contributed on behalf of STMicroelectronics

planger commented on packages/protocol/src/action-protocol/element-creation.ts:
```diff
@@ -14,69 +14,159 @@
+     */
     elementTypeId: string;
+    /**
+     * Optional additional arguments that might be needed by the server to execute the create operation.
```
```suggestion
     * Optional additional arguments for the server to execute the create operation.
```

planger commented on packages/protocol/src/action-protocol/element-hover.ts:
```diff
@@ -13,47 +13,79 @@
-    static create(elementId: string, bounds: Bounds): RequestAction<SetPopupModelAction> {
-        return new RequestPopupModelAction(elementId, bounds, generateRequestId());
-    }
+    /**
+     * The bounds.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/element-hover.ts:
```diff
@@ -13,47 +13,79 @@
-    static readonly KIND = 'setPopupModel';
-    readonly kind = SetPopupModelAction.KIND;
+export interface SetPopupModelAction extends ResponseAction, sprotty.SetPopupModelAction {
+    /**
+     * The unique action kind.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on examples/workflow-glsp/src/direct-task-editing/direct-task-editor.ts:
```diff
@@ -20,28 +20,60 @@ import {
+ * to execute a task edit operation.
+ */
+export interface ApplyTaskEditOperation extends Operation {
+    /**
+     * The unique action kind
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/client/src/base/actions/focus-change-action.ts:
```diff
@@ -13,14 +13,34 @@
+ * diagram focus changes (i.e. focus loss or focus gain).
+ */
+export interface FocusStateChangedAction extends Action {
+    /**
+     * The unique action kind
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/base-protocol.ts:
```diff
@@ -13,114 +13,224 @@
-
-    constructor(public readonly message: string, public readonly responseId: string, public readonly detail?: JsonAny) {}
+export interface RejectAction extends ResponseAction, sprotty.RejectAction {
+    /**
+     * The unique action kind.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/base-protocol.ts:
```diff
@@ -13,114 +13,224 @@
-    static readonly KIND = 'compound';
-    constructor(public operationList: Operation[], public readonly kind: string = CompoundOperation.KIND) {}
+export interface CompoundOperation extends Operation {
+    /**
+     * The unique action kind.
+     */
+    kind: typeof CompoundOperation.KIND;
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/client-notification.ts:
```diff
@@ -13,53 +13,113 @@
-        readonly kind = GLSPServerStatusAction.KIND
-    ) {}
+export interface ServerStatusAction extends Action {
+    /**
+     * The unique action kind.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/client-notification.ts:
```diff
@@ -13,53 +13,113 @@
+ * and creating new `ServerMessageActions`.
+ */
+export interface ServerMessageAction extends Action {
+    /**
+     * The unique action kind.
+     */
+    kind: typeof ServerMessageAction.KIND;
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/client-notification.ts:
```diff
@@ -13,53 +13,113 @@
+     */
+    kind: typeof ServerMessageAction.KIND;
+
+    /**
+     * The severity of the message.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/client-notification.ts:
```diff
@@ -13,53 +13,113 @@
+     */
+    severity: ServerSeverity;
+
+    /**
+     * The message text.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/clipboard.ts:
```diff
@@ -14,80 +14,154 @@
-export class RequestClipboardDataAction implements RequestAction<SetClipboardDataAction> {
-    static readonly KIND = 'requestClipboardData';
+export interface RequestClipboardDataAction extends RequestAction<SetClipboardDataAction> {
+    /**
+     * The unique action kind.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/clipboard.ts:
```diff
@@ -14,80 +14,154 @@
-        public readonly requestId: string = generateRequestId(),
-        public readonly kind: string = RequestClipboardDataAction.KIND
-    ) {}
+    /**
+     * The current editor context.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/clipboard.ts:
```diff
@@ -14,80 +14,154 @@
-        public readonly kind: string = SetClipboardDataAction.KIND
-    ) {}
+export interface SetClipboardDataAction extends ResponseAction {
+    /**
+     * The unique action kind.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/clipboard.ts:
```diff
@@ -14,80 +14,154 @@
+     */
+    kind: typeof SetClipboardDataAction.KIND;
+
+    /**
+     * The selected elements from the editor context as clipboard data.
+     */
```
```suggestion
    /**
     * The data to be added into the clipboard. This data will be sent back to the server on paste.
     */
```

planger commented on packages/protocol/src/action-protocol/clipboard.ts:
```diff
@@ -14,80 +14,154 @@
-export class CutOperation implements Operation {
-    static readonly KIND = 'cut';
+export interface CutOperation extends Operation {
+    /**
+     * The unique action kind.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/clipboard.ts:
```diff
@@ -14,80 +14,154 @@
+    kind: typeof CutOperation.KIND;
 
-    constructor(public readonly editorContext: EditorContext, public readonly kind: string = CutOperation.KIND) {}
+    /**
+     * The current editor context.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/clipboard.ts:
```diff
@@ -14,80 +14,154 @@
-    ) {}
-}
+export interface PasteOperation extends Operation {
+    /**
+     * The unique action kind.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/clipboard.ts:
```diff
@@ -14,80 +14,154 @@
 
-export function isPasteOperation(action: any): action is PasteOperation {
-    return isActionKind(action, PasteOperation.KIND) && isObject(action, 'clipboardData') && isObject(action, 'editorContext');
+    /**
+     * The current editor context.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/contexts.ts:
```diff
@@ -14,45 +14,83 @@
-        public readonly kind: string = RequestContextActions.KIND
-    ) {}
+export interface RequestContextActions extends RequestAction<SetContextActions> {
+    /**
+     * The unique action kind.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/contexts.ts:
```diff
@@ -14,45 +14,83 @@
+     */
+    contextId: string;
+
+    /**
+     * The current editor context.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

tortmayr replied:
I'm fine with omitting comments for self explaining primitive properties.
However, for object properties I think it would be beneficial to add a comment which links to the type of the property.
This way a user can just ctrl+ click to the definition if something is not clear.

planger commented on packages/protocol/src/action-protocol/contexts.ts:
```diff
@@ -14,45 +14,83 @@
-        public readonly kind: string = SetContextActions.KIND
-    ) {}
+export interface SetContextActions extends ResponseAction {
+    /**
+     * The unique action kind.
+     */
+    kind: typeof SetContextActions.KIND;
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/edge-modification.ts:
```diff
@@ -14,41 +14,98 @@
-        public readonly kind: string = ReconnectEdgeOperation.KIND
-    ) {}
+export interface ReconnectEdgeOperation extends Operation {
+    /**
+     * The unique action kind.
+     */
+    kind: typeof ReconnectEdgeOperation.KIND;
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/edge-modification.ts:
```diff
@@ -14,41 +14,98 @@
-    static readonly KIND = 'changeRoutingPoints';
-    constructor(public newRoutingPoints: ElementAndRoutingPoints[], public readonly kind: string = ChangeRoutingPointsOperation.KIND) {}
+export interface ChangeRoutingPointsOperation extends Operation {
+    /**
+     * The unique action kind.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/element-creation.ts:
```diff
@@ -14,69 +14,159 @@
-    ) {}
-}
+export interface CreateNodeOperation extends CreateOperation {
+    /**
+     * The unique action kind.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/element-creation.ts:
```diff
@@ -14,69 +14,159 @@
-    ) {}
-}
+export interface CreateEdgeOperation extends CreateOperation {
+    /**
+     * The unique action kind.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/element-creation.ts:
```diff
@@ -14,69 +14,159 @@
+     */
+    kind: typeof CreateEdgeOperation.KIND;
+
+    /*
+     * The source element.
+     */
```
```suggestion
    /*
     * The id of the source element of the edge to be created.
     */
```

planger commented on packages/protocol/src/action-protocol/element-creation.ts:
```diff
@@ -14,69 +14,159 @@
-        isString(action, 'sourceElementId') &&
-        isString(action, 'targetElementId')
-    );
+    /*
+     * The target element.
+     */
```
```suggestion
    /*
     * The id of the target element of the edge to be created.
     */
```

planger commented on packages/protocol/src/action-protocol/element-creation.ts:
```diff
@@ -14,69 +14,159 @@
-    static readonly KIND = 'deleteElement';
-    constructor(readonly elementIds: string[], public readonly kind: string = DeleteElementOperation.KIND) {}
+export interface DeleteElementOperation extends Operation {
+    /**
+     * The unique action kind.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/element-creation.ts:
```diff
@@ -14,69 +14,159 @@
+     */
+    kind: typeof DeleteElementOperation.KIND;
+
+    /**
+     * The elements to be deleted.
+     */
```
```suggestion
    /**
     * The ids of the elements to be deleted.
     */
```

planger commented on packages/protocol/src/action-protocol/element-hover.ts:
```diff
@@ -13,47 +13,79 @@
-    static readonly KIND = 'requestPopupModel';
-    readonly kind = RequestPopupModelAction.KIND;
+export interface RequestPopupModelAction extends RequestAction<SetPopupModelAction>, sprotty.RequestPopupModelAction {
+    /**
+     * The unique action kind.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.

planger commented on packages/protocol/src/action-protocol/element-navigation.ts:
```diff
@@ -13,56 +13,123 @@
+     * Denotes the position of the cursor in a text element.
+     */
     export interface TextPosition {
+        /**
+         * The current line number.
+         */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/element-navigation.ts:
```diff
@@ -13,56 +13,123 @@
+         * The current line number.
+         */
         line: number;
+        /**
+         * The character number where the cursor is currently located.
+         */
```
```suggestion
        /**
         * The character number within the line.
         */
```

planger commented on packages/protocol/src/action-protocol/element-navigation.ts:
```diff
@@ -86,91 +160,226 @@ export namespace NavigationTarget {
-    kind = RequestNavigationTargetsAction.KIND;
-    constructor(readonly targetTypeId: string, readonly editorContext: EditorContext, readonly requestId: string = generateRequestId()) {}
+export interface RequestNavigationTargetsAction extends RequestAction<SetNavigationTargetsAction> {
+    /**
+     * The unique action kind.
+     */
```
I'd spare such comments, also if it hurts to be "incomplete". But imho they just add noise and add no information.
```suggestion
```

planger commented on packages/protocol/src/action-protocol/element-navigation.ts:
```diff
@@ -86,91 +160,226 @@ export namespace NavigationTarget {
-    kind = SetNavigationTargetsAction.KIND;
-    constructor(readonly targets: NavigationTarget[], readonly responseId: string = '', readonly args?: Args) {}
+export interface SetNavigationTargetsAction extends ResponseAction {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/element-navigation.ts:
```diff
@@ -86,91 +160,226 @@ export namespace NavigationTarget {
+     */
+    kind: typeof SetNavigationTargetsAction.KIND;
+
+    /**
+     * A list of navigation targets.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/element-navigation.ts:
```diff
@@ -86,91 +160,226 @@ export namespace NavigationTarget {
+ * and creating new `NavigateToTargetActions`.
+ */
+export interface NavigateToTargetAction extends Action {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/element-navigation.ts:
```diff
@@ -86,91 +160,226 @@ export namespace NavigationTarget {
+     */
+    kind: typeof NavigateToTargetAction.KIND;
+
+    /**
+     * The target to which we navigate.
+     */
```
```suggestion
    /**
     * The target to which the client shall navigate.
     */
```

planger commented on packages/protocol/src/action-protocol/element-navigation.ts:
```diff
@@ -86,91 +160,226 @@ export namespace NavigationTarget {
-    kind = SetResolvedNavigationTargetAction.KIND;
-    constructor(readonly elementIds: string[] = [], readonly args?: Args, readonly responseId: string = '') {}
+export interface SetResolvedNavigationTargetAction extends ResponseAction {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/element-navigation.ts:
```diff
@@ -86,91 +160,226 @@ export namespace NavigationTarget {
-    readonly kind = NavigateToExternalTargetAction.KIND;
-    constructor(readonly target: NavigationTarget) {}
+export interface NavigateToExternalTargetAction extends Action {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/element-navigation.ts:
```diff
@@ -86,91 +160,226 @@ export namespace NavigationTarget {
+     */
+    kind: typeof NavigateToExternalTargetAction.KIND;
+
+    /**
+     * The target to which we navigate.
+     */
```
```suggestion
    /**
     * The diagram-external target to which the client shall navigate.
     */
```

planger commented on packages/protocol/src/action-protocol/element-selection.ts:
```diff
@@ -13,38 +13,79 @@
-    static readonly KIND = 'elementSelected';
-    kind = SelectAction.KIND;
+export interface SelectAction extends Action, sprotty.SelectAction {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/element-selection.ts:
```diff
@@ -13,38 +13,79 @@
-    static readonly KIND = 'allSelected';
-    kind = SelectAllAction.KIND;
+export interface SelectAllAction extends Action, sprotty.SelectAllAction {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/element-text-editing.ts:
```diff
@@ -13,96 +13,170 @@
+     * The unique action kind.
+     */
+    kind: typeof RequestEditValidationAction.KIND;
+    /**
```
```suggestion

    /**
```

planger commented on packages/protocol/src/action-protocol/element-text-editing.ts:
```diff
@@ -13,96 +13,170 @@
+     */
+    modelElementId: string;
+
+    /**
+     * Text that should be considered for the model element.
+     */
```
```suggestion
    /**
     * Text that should be validated for the given context and model element.
     */
```

planger commented on packages/protocol/src/action-protocol/element-text-editing.ts:
```diff
@@ -13,96 +13,170 @@
-export function isSetEditValidationResultAction(action: Action): action is SetEditValidationResultAction {
-    return isActionKind(action, SetEditValidationResultAction.KIND) && isObject(action, 'status') && isString(action, 'responseId');
-}
+    /**
+     * Validation status.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/element-type-hints.ts:
```diff
@@ -72,29 +82,70 @@ export interface EdgeTypeHint extends TypeHint {
-    static readonly KIND = 'requestTypeHints';
-    constructor(public readonly requestId = '', public readonly kind: string = RequestTypeHintsAction.KIND) {}
+export interface RequestTypeHintsAction extends RequestAction<SetTypeHintsAction> {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/element-type-hints.ts:
```diff
@@ -72,29 +82,70 @@ export interface EdgeTypeHint extends TypeHint {
+     */
+    kind: typeof SetTypeHintsAction.KIND;
+
+    /**
+     * The hints for shape types.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/element-validation.ts:
```diff
@@ -48,41 +49,100 @@ export namespace MarkerKind {
+     */
+    kind: typeof SetMarkersAction.KIND;
+
+    /**
+     * The list of markers that has been requested by the `RequestMarkersAction`.
+     */
```
```suggestion
    /**
     * The list of markers to be added to the diagram.
     */
```

planger commented on packages/protocol/src/action-protocol/element-validation.ts:
```diff
@@ -48,41 +49,100 @@ export namespace MarkerKind {
 
 /**
- * Response to the {@link RequestMarkersAction} containing all validation markers. Sent from the server to the client.
+ * Response to the {@link RequestMarkersAction} containing all validation markers. Typically sent from the server to the client.
```
```suggestion
  * Instructs the client to add markers to the diagram.
  * Typically, this is a response to the {@link RequestMarkersAction} containing all validation markers, but can be sent by the server at any time.
```

planger commented on packages/protocol/src/action-protocol/element-validation.ts:
```diff
@@ -48,41 +49,100 @@ export namespace MarkerKind {
-    static readonly KIND = 'deleteMarkers';
-    constructor(public readonly markers: Marker[], public readonly kind = DeleteMarkersAction.KIND) {}
+export interface DeleteMarkersAction extends Action {
+    /**
+     * The unique action kind.
+     */
+    kind: typeof DeleteMarkersAction.KIND;
```
```suggestion
    kind: typeof DeleteMarkersAction.KIND;
```

planger commented on packages/protocol/src/action-protocol/model-data.ts:
```diff
@@ -13,64 +13,145 @@
-    static readonly KIND = 'requestModel';
-    readonly kind = RequestModelAction.KIND;
+export interface RequestModelAction extends RequestAction<SetModelAction>, sprotty.RequestModelAction {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/model-data.ts:
```diff
@@ -13,64 +13,145 @@
-
-    constructor(public readonly newRoot: SModelRootSchema, public readonly animate: boolean = true) {}
+export interface UpdateModelAction extends Action, Omit<sprotty.UpdateModelAction, 'matches' | 'cause'> {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/model-data.ts:
```diff
@@ -13,64 +13,145 @@
+     * The unique action kind.
+     */
+    kind: typeof UpdateModelAction.KIND;
+    /**
+     * The new updates model root
+     */
```
```suggestion
    /**
     * The updated model root to set in the client.
     */
```

planger commented on packages/protocol/src/action-protocol/model-data.ts:
```diff
@@ -13,64 +13,145 @@
+ * and creating new `ModelSourceChangedActions`.
+ */
+export interface ModelSourceChangedAction extends Action {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/model-edit-mode.ts:
```diff
@@ -13,21 +13,45 @@
+ * and creating new `SetEditModeActions`.
+ */
+export interface SetEditModeAction extends Action {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/model-layout.ts:
```diff
@@ -0,0 +1,139 @@
+import { SModelRootSchema } from './model-structure';
+import { ElementAndAlignment, ElementAndBounds } from './types';
+
+/** Sent from the server to the client to request bounds for the given model. The model is rendered invisibly so the bounds can
+ * derived from the DOM. The response is a ComputedBoundsAction. This hidden rendering round-trip is necessary if the client is responsible
```
```suggestion
/**
 * Sent from the server to the client to request bounds for the given model. The model is rendered invisibly so the bounds can
 * derived from the DOM. The response is a ComputedBoundsAction. This hidden rendering round-trip is necessary if the client is responsible
```

planger commented on packages/protocol/src/action-protocol/model-layout.ts:
```diff
@@ -0,0 +1,139 @@
+ * and creating new `RequestBoundsActions`.
+ */
+export interface RequestBoundsAction extends RequestAction<ComputedBoundsAction>, sprotty.RequestBoundsAction {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/model-layout.ts:
```diff
@@ -0,0 +1,139 @@
+    kind: typeof RequestBoundsAction.KIND;
+
+    /**
+     * The model elements to consider to compute the new bounds.
```
```suggestion
     * The model root element for which to compute the new bounds.
```

planger commented on packages/protocol/src/action-protocol/model-layout.ts:
```diff
@@ -0,0 +1,139 @@
+ * and creating new `LayoutOperations`.
+ */
+export interface LayoutOperation extends Operation, Omit<sprotty.LayoutAction, 'layoutType'> {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/model-saving.ts:
```diff
@@ -13,42 +13,127 @@
+ * and creating new `SetDirtyStateActions`.
+ */
+export interface SetDirtyStateAction extends Action {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/model-saving.ts:
```diff
@@ -13,42 +13,127 @@
-    static readonly KIND = 'setDirtyState';
-    constructor(public readonly isDirty: boolean, public readonly reason?: string, public readonly kind = SetDirtyStateAction.KIND) {}
+/**
+ * The server sends a SetDirtyStateAction to indicate to the client that the current model state on the server does not correspond
```
```suggestion
 * The server sends this action to indicate to the client that the current model state on the server does not correspond
```

planger commented on packages/protocol/src/action-protocol/model-saving.ts:
```diff
@@ -13,42 +13,127 @@
+    isDirty: boolean;
+
+    /**
+     * A string indicating the reason for the dirty state change e.g 'operation', 'undo' ...
```
```suggestion
     * A string indicating the reason for the dirty state change e.g 'operation', 'undo', ...
```

planger commented on packages/protocol/src/action-protocol/model-structure.ts:
```diff
@@ -13,33 +13,57 @@
-export interface SModelElementSchema {
-    type: string;
+export interface SModelElementSchema extends sprotty.SModelElement {
+    /**
+     * Unique identifier for this element.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/node-modification.ts:
```diff
@@ -14,37 +14,86 @@
+     */
+    kind: typeof ChangeBoundsOperation.KIND;
+
+    /**
+     * The new bounds of the respective elements.
+     */
```
```suggestion
```

planger commented on packages/protocol/src/action-protocol/tool-palette.ts:
```diff
@@ -14,54 +14,85 @@
-export class TriggerNodeCreationAction extends TriggerElementCreationAction {
-    static readonly KIND = 'triggerNodeCreation';
+export interface TriggerNodeCreationAction extends Action {
+    /**
+     * The unique action kind.
+     */
```
```suggestion
```

planger reviewed (changes requested):
Excellent work! Thank you very much for this comprehensive change!
Great that you also added lots of typedoc and unit tests! :100: 

I made a couple of edit suggestions and comments about the typedocs. While most of them are great, there are several which imho don't really add information. I tend to prefer omitting those, because they add noise without helping.

I tested with https://github.com/eclipse-glsp/glsp-theia-integration/pull/110 and observed a few regressions:
  * Task editing doesn't work: after selecting type or duration, the input disappears
  * Validation throws an error (`Missing handler for action {  kind: 'applyMarkers',...`)
  * Resize and move restrictors for overlaps don't work... I'm not sure though if this is a regression or didn't work in the first place
  * Open push command (in Theia) doesn't select the correct element (could be unrelated too)

Is there a particular reason why we really copy and not import and re-export all Sprotty actions? I think there was a reason, but I can't remember right now. Now reviewing the code it feels like a lot of maintenance overhead with the risk of getting out of sync?

Again, excellent work on this huge change! Thanks for your persistence!

tortmayr commented on examples/workflow-glsp/src/direct-task-editing/direct-task-editor.ts:
```diff
@@ -156,6 +190,6 @@ export class TaskEditor extends AbstractUIExtension {
 }
 
-function getTask(ids: string[], index: SModelIndex<SModelElement>): TaskNode[] {
+function getTask(ids: string[], index: ModelIndexImpl): TaskNode[] {
```
Yes, IModelIndex doesn't offer the  `all(): FluentIterable<SModelElement>` method which we use quite often.

tortmayr commented:
Thank you for your fast review. I'm going to address the feedback right away. 

> Resize and move restrictors for overlaps don't work... I'm not sure though if this is a regression or didn't work in the first place

This is related to eclipse-glsp/glsp/issues/394. Currently the movement restrictor is not enabled for the workflow example.

>Is there a particular reason why we really copy and not import and re-export all Sprotty actions? I think there was a reason, but I can't remember right now. Now reviewing the code it feels like a lot of maintenance overhead with the risk of getting out of sync?

I assume you are referring to the actions in the protocol package?

 - With a simple reexport we lose the possibility to apply GLSP-specific JS-DOC to the action definitions
 - The sprotty  protocol actions use a slightly different approach regarding their namespace functions. They only offer the constant KIND and a `create` function. And (for whatever reason) opted to keep the guard functions as separate `isSomeAction()` functions. In GLSP we prefer guard functions as part of the namespace (is()). So we would have to do some kind of module augmentation anyways. In addtion, the  `isSomeAction` guard functions of sprotty are only available for some actions.
 - With re declaring we have the possibility to extend the sprotty provided interfaces. This is something we for instance do in the `ServerStatusAction` where we add an additional `timeout` property.
 - The parameterize of sprottiy's create() functions is not consistent and quite confusing at times. With the namespace re declare we have the possibility to apply our own consistent pattern.
- With re declaring the protocol actions we have a consistent entry place for API documentation. Otherwise the doc would
 be scattered across sprotty-protocol and @eclipse-glsp/protocol. With the current approach we have all protocol action documented in one place which also makes it relatively easy to sync doc changes with the https://github.com/eclipse-glsp/glsp/blob/master/PROTOCOL.md file.
 
Since we extend from the sprotty interfaces I don't see any risks of getting out of sync. This ensure that we are always compliant to the sprotty definition. I would even argue that is is a better approach in terms of API stability. With a simple reexport  we would pick up breaking changes much later (if at all).

planger commented:
Ah, right, now I see. Thanks!

tortmayr commented on packages/protocol/src/action-protocol/__tests__/element-text-editing.spec.ts:
```diff
@@ -0,0 +1,133 @@
+
+describe('Element text editing actions', () => {
+    describe('RequestEditValidationAction', () => {
+        it('RequestEditValidationAction.is with valid action type', () => {
```
I have update the test names in the `base-protocol.spec.ts` file. Could you maybe have a look at them. Feel free to make suggestions if you are not happy with the names. Once we agree on the naming in this file I will apply the same pattern to the other test files.

tortmayr commented on packages/client/src/features/command-palette/di.config.ts:
```diff
@@ -13,11 +13,9 @@
 import { CommandPalette, CommandPaletteActionProviderRegistry, TYPES } from 'sprotty';
-
+import '../../../css/command-palette.css';
 import { GLSP_TYPES } from '../../base/types';
```
We now export a  unified TYPES namespace from "@eclipse-glsp/client".
I kept the GLSP_TYPES namespace as deprecated export to ease migration pains when we try to adopt the nightly version in customer projects.

tortmayr commented:
I have pushed an update that should address all the raised issues.
There are a lot of comments regading JS-DOC so I hope that I didn't miss any. I had a quick check and all comments regarding doc seem to be outdated now (meaning the source code at the location has changed -> I hopefully addressed the comment :wink: )

> Task editing doesn't work: after selecting type or duration, the input disappears
Should be fixed now. Was related to a incorrect implementation of `LabeledAction.is`. With fixed I mean its in the current state
as the master. Apparently something is broken here. At least for me selecting a suggestion has no effect on the actual model. We should look into that in a follow-up.

>Validation throws an error (Missing handler for action { kind: 'applyMarkers',...)
Fixed. Was related to an incorrect `KIND` property of the `ApplyMarkers` command.

>Resize and move restrictors for overlaps don't work... I'm not sure though if this is a regression or didn't work in the first place
As mentioned before, movement restrictors for the Workflow example are currently disabled.

> Open push command (in Theia) doesn't select the correct element (could be unrelated too)
Fixed. It seems like we introduced duplicate names when changing the Workflow example styling. I updated all example1.wf files.

planger reviewed (approved):
Excellent, thanks a lot!
I did a few inline suggestions that you should be able to directly apply, mostly to typedoc.
So the only thing left is imho to adjust the text in the other tests. But I already approve for you to spare another review cycle.

Thanks again for your persistence with this large and important change! Now we are consistent and up to date with Sprotty again, and it is simpler for adopters when it comes to importing classes and registering DI implementations! :100:

tortmayr commented on packages/client/tsconfig.json:
```diff
@@ -6,5 +6,6 @@
     "reactNamespace": "JSX",
     "baseUrl": "."
   },
+
```
This change is auto applied by prettier, so I cannot really change htis.

tortmayr commented on packages/protocol/src/action-protocol/__tests__/base-protocol.spec.ts:
```diff
@@ -0,0 +1,237 @@
+/********************************************************************************
```
I personally am not a fan of  the approach of co-locating test files with the source they test. I think this makes it really hard to navigate the file tree and especially for packages with a lot of source files this quickly becomes convoluted.
Due to our poor coverage up until now this was not really an issue but when implementing the unit tests for the `action-protocol` package I already noticed that is was hard to distinguish source files and test files that why I opted for a dedicated tests package.

Maybe we should further discuss this with the team and then agree on a consistent structure.

planger replied:
Yeah, for sure a matter of taste. While I prefer to have tests and source next to each other, I'm for sure not strongly for or against one of the two approaches. I'm just saying that we should do it consistently at least per package. So once there is a general decision in the team, we should apply it consistently.

tortmayr replied:
For now I deleted the _tests__ folder and moved the tests alongside the source

tortmayr commented:
@planger  Thank you for the detailled review. I have updated the JS-DOC according to your suggestions and applied the new naming scheme to all test cases.  So now we should be ready to merge
